### PR TITLE
bpf(): improve map and program compatibility

### PR DESCRIPTION
**Pull request by lmb**
Allow passing a name for maps and programs and allow retrieving it via BPF_OBJ_GET_INFO_BY_FD. This requires a bunch of glue code because the bpf_prog_info, etc. structs are not compatible with Linux.

### CODE REVIEW

File: libs/api/bpf_syscall.cpp [hunk 1]
```diff
@@ -69,6 +72,105 @@ template <typename T> class ExtensibleStruct
...
+static void
+convert_to_sys_map_info(sys_bpf_map_info_t* sys, const struct bpf_map_info* bpf);
+static void
+convert_to_prog_info(struct bpf_prog_info* bpf, const sys_bpf_prog_info_t* sys);
+static void
+convert_to_sys_prog_info(sys_bpf_prog_info_t* sys, const struct bpf_prog_info* bpf);
+static void
+convert_to_link_info(struct bpf_link_info* bpf, const sys_bpf_link_info_t* sys);
+static void
+convert_to_sys_link_info(sys_bpf_link_info_t* sys, const struct bpf_link_info* bpf);
```
Comment (dthaler) [inline libs/api/bpf_syscall.cpp:98]: Can we add SAL annotations to these prototypes?

File: libs/api/bpf_syscall.cpp [hunk 2]
```diff
@@ -262,3 +380,84 @@ bpf(int cmd, union bpf_attr* attr, unsigned int size)
...
+
+#define BPF_TO_SYS(field) sys->field = bpf->field
+#define BPF_TO_SYS_STR(field) strncpy_s(sys->field, sizeof(sys->field), bpf->field, _TRUNCATE)
+#define BPF_TO_SYS_MEM(field) memcpy(sys->field, bpf->field, sizeof(sys->field))
+#define SYS_TO_BPF(field) bpf->field = sys->field
+#define SYS_TO_BPF_STR(field) strncpy_s(bpf->field, sys->field, sizeof(bpf->field))
+#define SYS_TO_BPF_MEM(field) memcpy(bpf->field, sys->field, sizeof(bpf->field))
+
+static void
+convert_to_map_info(struct bpf_map_info* bpf, const sys_bpf_map_info_t* sys)
```
Comment (dthaler) [inline libs/api/bpf_syscall.cpp:392]:
```suggestion
convert_to_map_info(_Out_ struct bpf_map_info* bpf, _In_ const sys_bpf_map_info_t* sys)
```

File: libs/api/bpf_syscall.cpp [hunk 3]
```diff
@@ -262,3 +380,84 @@ bpf(int cmd, union bpf_attr* attr, unsigned int size)
...
+{
+    SYS_TO_BPF(type);
+    SYS_TO_BPF(id);
+    // SYS_TO_BPF_MEM(tag);
+    // SYS_TO_BPF(jited_prog_len);
+    // SYS_TO_BPF(xlated_prog_len);
+    // SYS_TO_BPF(jited_prog_insns);
+    // SYS_TO_BPF(xlated_prog_insns);
+    // SYS_TO_BPF(load_time);
+    // SYS_TO_BPF(created_by_uid);
```
Comment (dthaler) [inline libs/api/bpf_syscall.cpp:426]: remove commented out code?

https://github.com/microsoft/ebpf-for-windows/blob/main/docs/DevelopmentGuide.md#coding-conventions says:
> **DON'T** use commented-out code, or code in an #if 0 or equivalent. Make sure all code is actually built.

File: tests/unit/libbpf_test.cpp [hunk 4]
```diff
@@ -3235,6 +3240,99 @@ TEST_CASE("BPF_MAP_GET_NEXT_KEY etc.", "[libbpf]")
...
+    REQUIRE(map_info.type == map_create.map_type);
+    REQUIRE(map_info.id != 0);
+    REQUIRE(map_info.key_size == map_create.key_size);
+    REQUIRE(map_info.value_size == map_create.value_size);
+    REQUIRE(map_info.max_entries == map_create.max_entries);
+    REQUIRE(map_info.map_flags == map_create.map_flags);
+    REQUIRE(strncmp(map_info.name, map_create.map_name, sizeof(map_info.name)) == 0);
+
+    struct ebpf_inst instructions[] = {
+        {0xb7, R0_RETURN_VALUE, 0}, // r0 = 0
```
Comment (dthaler) [inline tests/unit/libbpf_test.cpp:3286]:
```suggestion
        {INST_ALU_OP_MOV | INST_CLS_ALU64, R0_RETURN_VALUE, 0}, // r0 = 0
```

**Review by dthaler — APPROVED** [review]

**Comment by shankarseal** [comment]
@lmb - the CICD tests are failing on this PR. Can you please take a look?

**Comment by lmb** [comment]
Yes, I’m aware. My dev environment destroyed itself and setting it back up had to take lower priority. I’ll update when I can.

**Comment by lmb** [comment]
Rebased this on `main`. Compilation fails again for some unfathomable reason so I can't test this locally...

**Comment by saxena-anurag** [comment]
> Rebased this on `main`. Compilation fails again for some unfathomable reason so I can't test this locally...

Hi, since the build is passing in CICD pipeline, you can probably use the artifacts from the pipeline to run the tests locally.

**Comment by lmb** [comment]
Tests are finally passing!

File: libs/execution_context/ebpf_link.c [hunk 5]
```diff
@@ -652,7 +652,7 @@ ebpf_link_get_info(
 
     memset(info, 0, sizeof(*info));
     info->id = link->object.id;
-    info->prog_id = (link->program) ? ((ebpf_core_object_t*)link->program)->id : EBPF_ID_NONE;
+    info->prog_id = (link->program) ? ((ebpf_core_object_t*)link->program)->id : 0;
```
Comment (dthaler) [inline libs/execution_context/ebpf_link.c:655]: Question: should EBPF_ID_NONE be defined to be 0?

File: tests/unit/libbpf_test.cpp [hunk 6]
```diff
@@ -2259,6 +2263,19 @@ TEST_CASE("enumerate link IDs with bpf", "[libbpf]")
...
     REQUIRE(info.id == id1);
 
+    // Get info on the second link.
+    memset(&attr, 0, sizeof(attr));
+    info = {};
+    attr.info.bpf_fd = fd2;
+    attr.info.info = (uintptr_t)&info;
+    attr.info.info_len = sizeof(info);
+    REQUIRE(bpf(BPF_OBJ_GET_INFO_BY_FD, &attr, sizeof(attr)) == 0);
+    // TODO: Should this be BPF_LINK_TYPE_PLAIN?
```
Comment (dthaler) [inline tests/unit/libbpf_test.cpp:2273]: As far as I know, yes.
```suggestion
    // TODO: This should be BPF_LINK_TYPE_PLAIN.
```

**Review by dthaler — DISMISSED** [review]

File: tests/unit/libbpf_test.cpp (hunk 6 above)
Comment (lmb) [reply tests/unit/libbpf_test.cpp:2273]: See https://github.com/microsoft/ebpf-for-windows/issues/4096

File: libs/execution_context/ebpf_link.c (hunk 5 above)
Comment (lmb) [reply libs/execution_context/ebpf_link.c:655]: https://github.com/microsoft/ebpf-for-windows/issues/4098

**Review by dthaler — APPROVED** [review]

**Review by Alan-Jowett — APPROVED** [review]

**Pull request merged**